CMakeLists.txt

```diff
@@ -0,0 +1,172 @@
+cmake_minimum_required(VERSION 2.6)
+
+project(UriParser C)
+
+set(VERSION_STRING "0.8.6")
+IF(EXISTS "${CMAKE_SOURCE_DIR}/.git")
+    FIND_PROGRAM(GIT git)
+    IF(GIT)
+        EXECUTE_PROCESS(
+            WORKING_DIRECTORY ${CMAKE_SOURCE_DIR}
+            COMMAND ${GIT} describe --tags
+            OUTPUT_VARIABLE GIT_VERSION  OUTPUT_STRIP_TRAILING_WHITESPACE
+        )
+        if(NOT VERSION_STRING)
+                EXECUTE_PROCESS(
+                    WORKING_DIRECTORY ${CMAKE_SOURCE_DIR}
+                    COMMAND ${GIT} rev-parse --short HEAD
+                    OUTPUT_VARIABLE GIT_VERSION  OUTPUT_STRIP_TRAILING_WHITESPACE
+                )
+        endif()
+        SET(VERSION_STRING ${GIT_VERSION})
+    ENDIF(GIT)
+ENDIF()
+MESSAGE("Version: ${VERSION_STRING}\n")
+set(URIPARSER_VERSION ${VERSION_STRING})
+
+option(BUILD_TESTS "Build demos." OFF)
+
+IF(MSVC)
+    # This option is to enable the /MP switch for Visual Studio 2005 and above compilers
+    OPTION(WIN32_USE_MP "Set to ON to build OpenSceneGraph with the /MP option (Visual Studio 2005 and above)." ON)
+    MARK_AS_ADVANCED(WIN32_USE_MP)
+    IF(WIN32_USE_MP)
+        SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /MP")
+    ENDIF(WIN32_USE_MP)
+ENDIF(MSVC)
+
+include(CMakePackageConfigHelpers)
+include(GNUInstallDirs OPTIONAL)
+include(GenerateExportHeader)
+include(CheckIncludeFile)
+include(CheckFunctionExists)
+
+check_include_file("wchar.h" HAVE_WCHAR_H)
+check_include_file("stdio.h" HAVE_STDIO_H)

```

Comment:
I had a look for use of system (and third-party) headers. This is the full list I found, categorized by when (or what for) they are needed:
```
library:
    assert.h
    config.h
    ctype.h
    errno.h
    limits.h
    stdio.h
    stdlib.h
    string.h

library with wide char support and/or test suite:
    wchar.h

when building tests:
    gtest/gtest.h  # Googletest >=1.8.1

when building CLI tool:
    for windows:
        winsock2.h
        ws2tcpip.h

    for non-windows:
        sys/socket.h
        arpa/inet.h
        netinet/in.h
```

I'm unsure if we need to check for all of them. What do you think?

Reply:
when building tests:
    gtest/gtest.h  # Googletest >=1.8.1

To rely on gtest, please add after the merger

Reply:
> To rely on gtest, please add after the merger

I don't understand, please elaborate.